Sources/StreamChatUI/ChatMessageList/ChatMessageListView.swift

```diff
@@ -157,101 +157,34 @@ open class ChatMessageListView: UITableView, Customizable, ComponentsProvider {
         with changes: [ListChange<ChatMessage>],
         completion: (() -> Void)? = nil
     ) {
-        defer {
-            let lastMessageInserted = changes.first(where: { $0.isInsertion && $0.indexPath.row == 0 })?.item
+        let lastMessageInserted = changes.first(where: { $0.isInsertion && $0.indexPath.row == 0 })?.item
+        let scrollToBottomIfNeeded = {
             if lastMessageInserted?.isSentByCurrentUser == true {
-                scrollToMostRecentMessage()
+                self.scrollToMostRecentMessage()
             }
-
-            completion?()
         }
 
-        let hasConflictUpdates = collectionUpdatesMapper.mapToSetsOfIndexPaths(changes: changes) == nil
-        if hasConflictUpdates {
-            reloadData()
-            return
-        }
-
-        guard changes.count == 1, let change = changes.first else {
+        guard let indices = collectionUpdatesMapper.mapToSetsOfIndexPaths(
+            changes: changes
+        ) else {
```

Comment:
What if we move `collectionUpdatesMapper` to list database observer and in both UI frameworks just blindly apply what comes from LLC?

Reply:
We can't do that, because we need to notify about `ListChange<T>`, and the `collectionUpdatesMapper` returns index paths only. But I thought about introducing a `TableViewSafeBatchUpdater` and `CollectionViewBatchUpdater` which would be useful to reuse this logic between all table views and collection views. 👍 Will notify you once I change this 👍

Reply:
What if it returns `[ListChange<T>]` but with conflicting updates removed?

Reply:
It will be 2x the work. Because then we will need to stilll map them and separate Inserts, Updates, Deletes and Moves, and call them in the correct order. So it is better we do this in the same for-loop to improve performance

Reply:
We say that customers can use LLC without a UI framework but in fact, it's not possible if LLC keeps publishing crash-prone updates.

We indeed will have an increase in time complexity for additional changes processing but it'll be a constant one (`2*O(N)` instead of `O(N)`). The `N` of changes is relatively small while the benefit would be a fully functional LLC and no code duplication of patches in message/channel lists

Reply:
I'm ok to move it to tech discussion, no need to block a PR with that 👍

Reply:
@evsaev The PR is not blocked, I just didn't have the time yet to pick it up again. Will do it in the next couple of days. And yes I was planning on exposing this to customers and writing documentation on how they can use this with their own UI 🙂 Let me work on a WIP implementation of this first, and then we can discuss it together and bring it to tech discussion 👍

Reply:
@evsaev Added the implementation with the concept of `ListChangeUpdater`. IMO this is a pretty solid solution, but let's discuss this in the Tech Meeting 👍 The problem with your initial idea is that it doesn't solve the fact that we need to make sure `performBatchUpdates()` is called in the correct order. With this solution, the updates to collection/table views are abstracted from the customer and they can use these updaters easily with their custom views 🙂